src/go/rpk/pkg/config/v21_1_4.go

```diff
@@ -91,12 +91,15 @@ func Decodev2114(m map[string]interface{}) (*Config, error) {
 			from, to reflect.Type,
 			data interface{},
 		) (interface{}, error) {
-			if to == reflect.TypeOf(&v2114SocketAddress{}) {
+			if to == reflect.TypeOf([]v2114SocketAddress{}) {
```

Comment:
This make sense to me.  The implicit acceptable alternative to `reflect.Map` is `reflect.SliceOf` - I think it's ok to be implicit here, because the default Decode behaviour will do the right thing whether the conversion works or not.

I don't expect there to be a lot of overrides here, but I wonder if it's worth having a pattern, specified by an interface. I propose something along the lines of the error interface. It could extended it with validation later.

```go
type Transformer interface {
	Is(to reflect.Type) bool
	As(from reflect.Type, data interface{}) (interface{}, error)
}

type v2114SocketAddresses struct {
	val []v2114NamedSocketAddress
}

func (sas v2114SocketAddresses) Is(to reflect.Type) bool {
	return to == reflect.TypeOf(sas.val)
}

func (sas v2114SocketAddresses) As(from reflect.Type, data interface{}) (interface{}, error) {
		switch from.Kind() {
		case reflect.Map:
			sa := v2114NamedSocketAddress{}
			err := mapstructure.Decode(data, &sa)
			if err != nil {
				return nil, err
			}
			return []v2114NamedSocketAddress{sa}, nil

		}
	return data, nil
}

func Is(val interface{}, to reflect.Type) bool {
	v, ok := val.(Transformer);
	if !ok {
		return false
	}
	return v.Is(to)
}

func As(val interface{}, from reflect.Type, data interface{}) (interface{}, error) {
	v, ok := val.(Transformer);
	if !ok {
		return nil, errors.New("")
	}
	return v.As(from, data)
}


func Decodev2114(m map[string]interface{}) (*Config, error) {
	result := &v2114Config{}
	decoderConfig := mapstructure.DecoderConfig{
		Result:	result,
		DecodeHook: func(
			from, to reflect.Type,
			data interface{},
		) (interface{}, error) {
			if Is(&v2114SocketAddresses{}, to) {
				return As(&v2114SocketAddresses{}, from, data)
			}

			return data, nil
		},
	}
	decoder, err := mapstructure.NewDecoder(&decoderConfig)
	if err != nil {
		return nil, err
	}
	err = decoder.Decode(m)
	if err != nil {
		return nil, err
	}
	return result.ToGeneric()
}
```

That should make registration of new transformations easy, and those transformations the live with the type.

Ostensibly there's an `if {} else if {} else {}` chain, that could be further refactored to a wrapper over a list, in priority order.

Reply:
This looks good. There won't be many implementations right now, but it sets a precedent.

Reply:
Now I wonder if the Is/As split is the right thing. Maybe each Transformer should just implement [DecodeHookFunc](https://pkg.go.dev/github.com/mitchellh/mapstructure#DecodeHookFunc) directly, and do the passthough.